src/metabase/api/public.clj

```diff
@@ -51,21 +51,44 @@
 
 ;;; -------------------------------------------------- Public Cards --------------------------------------------------
 
+(defn combine-parameters-and-template-tags
+  "Update `card.parameters` to include parameters from template-tags.
+
+  On native queries parameters exists in 2 forms:
+  - parameters
+  - dataset_query.native.template-tags
+
+  In most cases, these 2 are sync, meaning, if you have a template-tag, there will be a parameter.
+  However, since card.parameters is a recently added feature, there may be instances where a template-tag
+  is not present in the parameters.
+  This function ensures that all template-tags are converted to parameters and added to card.parameters."
+  [{:keys [parameters] :as card}]
+  (let [template-tag-parameters     (card/template-tag-parameters card)
+        id->template-tags-parameter (m/index-by :id template-tag-parameters)
+        id->parameter               (m/index-by :id parameters)
+        all-parameter-ids           (set (concat (map :id template-tag-parameters)
+                                                 (map :id parameters)))
+        parameters                  (for [id all-parameter-ids]
+                                      (merge (get id->parameter id {})
+                                             (get id->template-tags-parameter id {})))]
```

Comment:
Aesthetic suggestion:
```suggestion
(vals (reduce-kv (fn [acc id parameter]
                     (update acc id #(merge % parameter)))
                   id->parameter
                   id->template-tags-parameter))
```

Reply:
Why merge the template tags parameters on top of card.parameters, instead of overriding them completely, or doing it the other way around? Or does it not matter, because each key's values are guaranteed to be equal? It might be nice to explain this in the docstring.

Reply:
yep that is better. also I added the test to check order of merge.